src/systemcmds/bsondump/bsondump.cpp

```diff
@@ -82,58 +91,224 @@ static int bson_print_callback(bson_decoder_t decoder, bson_node_t node)
+{
+	int src_fd = open(src_path, O_RDONLY);
+	int dst_fd = open(dst_path, O_WRONLY | O_TRUNC);
+
+	if (src_fd == -1 || dst_fd == -1) {
+		perror("Failed to open files for copying");
+
+		if (src_fd != -1) { close(src_fd); }
+
+		if (dst_fd != -1) { close(dst_fd); }
+
+		return -1;
+	}
+
+	char buffer[1024];
+	ssize_t bytes_read;
+
+	while ((bytes_read = read(src_fd, buffer, sizeof(buffer))) > 0) {
+		if (write(dst_fd, buffer, bytes_read) != bytes_read) {
+			perror("Failed to copy file");
+			close(src_fd);
+			close(dst_fd);
+			return -1;
+		}
+	}
+
+	close(src_fd);
+	close(dst_fd);
+
+	if (bytes_read == -1) {
+		perror("Error reading from source file");
+		return -1;
+	}
+
+	return 0;
+}
+
 extern "C" __EXPORT int bsondump_main(int argc, char *argv[])
 {
-	if (argc != 2) {
-		print_usage();
+	if (argc < 2) {
+		print_usage("Invalid number of arguments.");
+		return -1;
+	}
+
+	char *command = argv[1];
+	char *file_name = argv[3];
+
+	if (strcmp(command, "docsize") == 0) {
+		if (argc != 3) {
+			print_usage("Usage: bsondump docsize <file>", "docsize");
+			return -1;
+		}
+
+		file_name = argv[2];
+		int source_fd = open(file_name, O_RDONLY);
+
+		if (source_fd < 0) {
+			PX4_ERR("open '%s' failed (%i)", file_name, errno);
+			return 1;
+		}
+
+		bson_decoder_s decoder{};
+		int result = -1;
+
+		if (bson_decoder_init_file(&decoder, source_fd, bson_docsize_callback) == 0) {
+
+			do {
+				result = bson_decoder_next(&decoder);
+
+			} while (result > 0);
+
+			PX4_INFO("DECODED_SIZE:%" PRId32 " SAVED_SIZE:%" PRId32 "\n",
+				 decoder.total_decoded_size, decoder.total_document_size);
+		}
+
+		close(source_fd);
+
+		if (decoder.total_decoded_size != decoder.total_document_size && decoder.total_document_size == 0
+		    && strstr(file_name, "mtd_caldata") != NULL) {
+
+			PX4_WARN("Mismatch in BSON sizes and saved size is zero. Setting document size to decoded size.");
+
+			const char *source_path = "/fs/mtd_caldata";
+			const char *temp_path = "/fs/microsd/tmp/mtd_caldata";
+
+			source_fd = open(source_path, O_RDONLY);
+
+			if (source_fd == -1) {
+				perror("Failed to re-open source file for reading");
+				return -1;
+			}
+
+			struct stat st;
+
+			if (fstat(source_fd, &st) != 0) {
+				perror("Failed to stat source file");
+				close(source_fd);
+				return -1;
+			}
+
+			char *buffer = new char[st.st_size];
+
+			if (read(source_fd, buffer, st.st_size) != st.st_size) {
+				perror("Failed to read source file");
+				delete[] buffer;
+				close(source_fd);
+				return -1;
+			}
+
+			// Modify the first 4 bytes with the correct decoded size
+			uint32_t corrected_size = decoder.total_decoded_size;
+			memcpy(buffer, &corrected_size, sizeof(corrected_size));
+
+			char *dir_path = strdup(temp_path);
+			char *dir = dirname(dir_path);
+
+			struct stat st2 = {0};
+
+			if (stat(dir, &st2) == -1) {
+				mkdir(dir, 0777);
+			}
+
+			free(dir_path);
+
+			int temp_fd = open(temp_path, O_WRONLY | O_CREAT | O_TRUNC, 0666);
```

Comment:
Why do you write to a temp file here instead of directly to source_path?
You also don't need to write the whole file, just the first 4 bytes.

Reply:
just to have a backup and maybe some checks before writing on src

Reply:
fixed